testing/ase/test_enum_c_ase.cpp

```diff
@@ -0,0 +1,380 @@
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+extern "C" {
+#include <json-c/json.h>
+#include <uuid/uuid.h>
+#include <opae/types.h>
+#include "types_int.h"
+#include "properties_int.h"
+#include "common_int.h"
+}
+
+#include <opae/fpga.h>
+
+#include <cstdlib>
+#include <memory>
+#include "gtest/gtest.h"
+
+// ASE ID
+#define ASE_TOKEN_MAGIC    0x46504741544f4b40
+
+static const fpga_guid ASE_GUID = {
+    0xd8, 0x42, 0x4d, 0xc4, 0xa4,  0xa3, 0xc4, 0x13, 0xf8,0x9e,
+    0x43, 0x36, 0x83, 0xf9, 0x04, 0x0b
+};
+
+inline void token_for_fme(struct _fpga_token* tok_)
+{
+    std::copy(&FPGA_FME_GUID[0], &FPGA_FME_GUID[16], tok_->accelerator_id);
+    tok_->magic = ASE_TOKEN_MAGIC;
+    tok_->ase_objtype = FPGA_DEVICE;
+}
+
+inline void token_for_afu0(struct _fpga_token* tok_)
+{
+    std::copy(&ASE_GUID[0], &ASE_GUID[16], tok_->accelerator_id);
+    tok_->magic = ASE_TOKEN_MAGIC;
+    tok_->ase_objtype = FPGA_ACCELERATOR;
+}
+
+fpga_guid known_guid = {0xc5, 0x14, 0x92, 0x82, 0xe3, 0x4f, 0x11, 0xe6,
+                        0x8e, 0x3a, 0x13, 0xcc, 0x9d, 0x38, 0xca, 0x28};
+
+class enum_c_ase_p : public testing::Test {
+  protected:
+    enum_c_ase_p() : tok(nullptr) {}
+
+    virtual void SetUp() override {
+        token_for_afu0(&tok_);
+        tok = &tok_;
+
+        filter_ = nullptr;
+        ASSERT_EQ(fpgaGetProperties(nullptr, &filter_), FPGA_OK);
+    }
+
+    virtual void TearDown() override {
+        if (filter_ != nullptr) {
+            EXPECT_EQ(fpgaDestroyProperties(&filter_), FPGA_OK);
+        }
+    }
+
+    fpga_properties filter_;
+    struct _fpga_token tok_;
+    fpga_token tok;
+};
+
+/**
+ * @test       nullfilter
+ *
+ * @brief      When the fpga_properties * parameter to fpgaEnumerate is nullptr, the
+ *             function returns FPGA_INVALID_PARAM.
+ */
+TEST_F(enum_c_ase_p, nullfilter) {
+  uint32_t matches = 0;
+  EXPECT_EQ(fpgaEnumerate(nullptr, 1, &tok, 1, &matches),
+            FPGA_INVALID_PARAM);
+}
+
+/**
+ * @test       nullmatches
+ *
+ * @brief      When the uint32_t *num_matches parameter to fpgaEnumerate is nullptr, the
+ *             function returns FPGA_INVALID_PARAM.
+ */
+TEST_F(enum_c_ase_p, nullmatches) {
+  uint32_t matches = 0;
+  EXPECT_EQ(fpgaEnumerate(&filter_, 1, &tok, 1, NULL),
+            FPGA_INVALID_PARAM);
+  EXPECT_EQ(
+      fpgaEnumerate(&filter_, 0, &tok, 1, &matches),
```

Comment:
could be wrong but this looks like it should return FPGA_OK?

Reply:
Agree.. why is this failing?

Reply:
When the filter is not NULL but the number of filter is zero, it will return FPGA_INVALID_PARAM.

Reply:
doesn't 0 mean match all FPGA resources?

Reply:
nvm, 0 to match all FPGA resources when filters is NULL. But filter isn't.

Reply:
At line 574 of libopae/api-shell.c, it said 	
        if ((num_filters > 0) && !filters) {
		OPAE_ERR("num_filters > 0 with NULL filters");
		return FPGA_INVALID_PARAM;
	}

Reply:
You're passing filter_ here, which I think is non-NULL.

Reply:
> When the filter is not NULL but the number of filter is zero, it will return FPGA_INVALID_PARAM.
Ok, I see it now.  Yes, agree.  Please update the test comments to reflect this point.

Reply:
Will update doc